### PR TITLE
fix[lang]: disallow absolute relative imports

### PR DESCRIPTION
**Pull request by sandbubbles**
### What I did
- fix https://github.com/vyperlang/vyper/issues/4184
### How I did it
- Removed implicit relative imports in submodules, eliminating ambiguity in import paths.
- Removed the lines that added the current directory to the top of the search paths. As a result, the poke_search_path function may no longer be necessary.

### How to verify it
- added tests

### Commit message

````
this commit changes the behavior of search paths within
submodules. previously the current module's parent was appended to
the search path in the import recursion. this means that given the
following directory structure, this code is legal:

```
# directory structure:
# subdir/foo.vy
# subdir/bar.vy
```

```
# subdir/foo.vy
import bar
```

which has the same semantics as

```
# subdir/foo.vy
from . import bar
```

this represented a divergence wrt python's import system. in python,
only the second example is allowed.

this commit changes the semantics to match python, so that relative
imports use the current directory as search path, but absolute imports
can only use the top-level search path. (the directory containing the
top-level compilation target is still included in the search path, but
this only happens at the top of the recursion, same as in python -
https://docs.python.org/3/library/sys_path_init.html).

misc/refactor:
- remove dead `add_search_path()` method
- remove now-dead `poke_search_path()`
````

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

**Review by charles-cooper — COMMENTED** [review]
yes, i think poke_search_paths is probably dead now. can you check? if so, let's remove it!

File: vyper/semantics/analysis/module.py [hunk 1]
```diff
@@ -789,11 +789,7 @@ def _add_import(
     # load an InterfaceT or ModuleInfo from an import.
     # raises FileNotFoundError
     def _load_import(self, node: vy_ast.VyperNode, level: int, module_str: str, alias: str) -> Any:
```
Comment (charles-cooper) [inline vyper/semantics/analysis/module.py:791]: i guess we can fuse load_import_helper right?

File: vyper/semantics/analysis/module.py [hunk 2]
```diff
@@ -789,7 +790,13 @@ def _add_import(
...
-        return self._load_import_helper(node, level, module_str, alias)
+        # the directory this (currently being analyzed) module is in
+        self_search_path = Path(self.ast.resolved_path).parent
+
+        with self.input_bundle.search_path(
```
Comment (charles-cooper) [inline vyper/semantics/analysis/module.py:796]: i think for a relative import, we actually need to override the entire search path. can be a context manager, but something along the lines of
```
try:
    self.input_bundle.search_paths = [self_search_path]
    self._load_import_helper(...)
finally:
    self.input_bundle.search_paths = original_search_paths
```

otherwise you can break out of the current directory if the search path falls back to some other path (e.g. like /usr/lib/site-packages/foo/bar)

Comment (sandbubbles) [reply vyper/semantics/analysis/module.py:795]: The proposed solution is similar - it's just tricky when we have a file with a relative path import that imports something with an absolute path. Inside that file, we need all the search paths (which would be deleted by the previous relative import). Therefore, the new solution keeps all the search paths and switches between searching only the topmost path and all the paths.

Comment (charles-cooper) [reply vyper/semantics/analysis/module.py:795]: Tricky indeed! I'm not sure I like changing the InputBundle api just for this, it seems like it should rather be part of the recursion state.

Comment (sandbubbles) [reply vyper/semantics/analysis/module.py:795]: True true!  and there is another problem with the solution now  - the "current search paths" are appended to the list, so an absolute import could look into a "current search path" of a previous relative import.

File: vyper/semantics/analysis/imports.py [hunk 3]
```diff
@@ -80,6 +80,9 @@ def __init__(self, input_bundle: InputBundle, graph: _ImportGraph):
         self.seen: set[int] = set()
 
         self.integrity_sum = None
+        self.abosulte_search_paths = (
```
Comment (charles-cooper) [inline vyper/semantics/analysis/imports.py:83]: typo

File: vyper/semantics/analysis/imports.py [hunk 4]
```diff
@@ -80,6 +80,9 @@ def __init__(self, input_bundle: InputBundle, graph: _ImportGraph):
...
+            input_bundle.search_paths.copy()
+        )  # should be all system paths + topmost module path
```
Comment (charles-cooper) [inline vyper/semantics/analysis/imports.py:85]: multiline expression
```suggestion
        # should be all system paths + topmost module path
        self.absolute_search_paths = input_bundle.search_paths.copy()
```

File: tests/unit/cli/vyper_json/test_compile_json.py [hunk 5]
```diff
@@ -318,4 +318,6 @@ def test_compile_json_with_abi_top(make_input_bundle):
 from . import stream
     """
     input_bundle = make_input_bundle({"stream.json": stream, "code.vy": code})
-    vyper.compiler.compile_code(code, input_bundle=input_bundle)
+    vyper.compiler.compile_code(
+        code, resolved_path=input_bundle.search_paths[0] / "code.vy", input_bundle=input_bundle
+    )
```
Comment (charles-cooper) [inline tests/unit/cli/vyper_json/test_compile_json.py:323]: why is this change needed?

Comment (sandbubbles) [reply tests/unit/cli/vyper_json/test_compile_json.py:323]: Because it imports by a relative path, it needs the current path of the file. By default it is left to unknown (`UNKNOWN_CONTRACT_NAME`), so then when we ask for the current directory (by `.parent) it gets evaluated to "." instead of the temporary directory. But i think you are right that i added it somewhere unnecessarily.

**Comment by cyberthirst** [comment]
@charles-cooper, have you considered the concept of packages? because we don't have it we allow relative imports which can point even outside the project's directory

File: tests/functional/codegen/test_interfaces.py [hunk 6]
```diff
@@ -171,7 +169,32 @@ def bar() -> uint256:
...
+
+
+# TODO CMC 2024-10-13: should probably be in syntax tests
+@pytest.mark.parametrize("code,subdirs", VALID_RELATIVE_IMPORT_CODE)
+def test_extract_file_interface_relative_imports(code, subdirs, make_input_bundle):
+    input_bundle = make_input_bundle({"a/Foo.vyi": ""})
+
+    assert (
+        compile_code(
+            code, resolved_path=(input_bundle.search_paths[0]) / subdirs, input_bundle=input_bundle
```
Comment (cyberthirst) [inline tests/functional/codegen/test_interfaces.py:194]: shouldn't the resolved path end with a file? like `code` for the 2nd case should be in `a/b`, right? in which case `..a` resolves to `a/a` and we should get file not found?

**Comment by charles-cooper** [comment]
> @charles-cooper, have you considered the concept of packages? because we don't have it, we allow relative imports that can point even outside the project's directory

yeah i know, those will raise a compiler panic if you try to create an input bundle.

i think it might be useful to introduce packages, but it needs some design work, and not sure how useful it would really be.

**Comment by cyberthirst** [comment]
> yeah i know, those will raise a compiler panic if you try to create an input bundle.

i have `main.vy` in `./tests/custom` and run vyper from the same dir as in which `tests` are

```vyper
from .... import lib1

exports: lib1.__interface__
```
so the import targets a file which is in the parent of the project's dir (from which vyper's run) .. and it compiles fine

File: vyper/semantics/analysis/imports.py [hunk 7]
```diff
@@ -149,17 +152,19 @@ def _add_import(
...
+        if level == 0:
+            self.input_bundle.search_paths = self.absolute_search_paths
+        else:
+            ast = self.graph.current_module
+            current_search_path = Path(ast.resolved_path).parent
+            self.input_bundle.search_paths = [current_search_path]
 
-        with self.input_bundle.poke_search_path(self_search_path):
-            return self._load_import_helper(node, level, module_str, alias)
+        return self.input_bundle.load_file(path)
```
Comment (charles-cooper) [inline vyper/semantics/analysis/imports.py:163]: nit: can we move this after `_load_import` to match the other functions, which are laid out in reverse dependency order

File: vyper/semantics/analysis/imports.py [hunk 8]
```diff
@@ -224,6 +214,16 @@ def _load_import(
         search_paths = self.input_bundle.search_paths.copy()  # noqa: F841
         raise ModuleNotFound(module_str, hint=hint) from err
 
+    def _load_file(self, path: PathLike, level: int):
+        if level == 0:
+            self.input_bundle.search_paths = self.absolute_search_paths
+        else:
+            ast = self.graph.current_module
+            current_search_path = Path(ast.resolved_path).parent
+            self.input_bundle.search_paths = [current_search_path]
```
Comment (charles-cooper) [inline vyper/semantics/analysis/imports.py:223]: i think this corrupts the input bundle's search path, which might be important if anybody tries to access the search path after compilation (for instance as is done in `output_bundle.py`). i think we need to restore the original `.search_paths` after `load_file` is called.

**Comment by cyberthirst** [comment]
> > yeah i know, those will raise a compiler panic if you try to create an input bundle.
> 
> i have `main.vy` in `./tests/custom` and run vyper from the same dir as in which `tests` are
> 
> ```vyper
> from .... import lib1
> 
> exports: lib1.__interface__
> ```
> 
> so the import targets a file which is in the parent of the project's dir (from which vyper's run) .. and it compiles fine

i tried to create an output bundle (vyper -f archive) and it leads to compiler panic (modified the path in the error so it doesn't show my data)

```
Error compiling: tests/custom/test4.vy
vyper.exceptions.CompilerPanic: Invalid path: /path/to_lib1/lib1.vy

This is an unhandled internal compiler error. Please create an issue on Github to notify the developers!
https://github.com/vyperlang/vyper/issues/new?template=bug.md
```

**Comment by charles-cooper** [comment]
@sandbubbles please resolve merge conflicts

File: vyper/semantics/analysis/imports.py [hunk 9]
```diff
@@ -219,6 +214,20 @@ def _load_import_helper(
...
+        ast = self.graph.current_module
+        current_search_path = Path(ast.resolved_path).parent
+
+        search_paths = self.absolute_search_paths if level == 0 else [current_search_path]
+
+        with self.input_bundle.temporary_search_paths(search_paths):
+            res = self.input_bundle.load_file(path)
+
+        if level != 0:
+            self.input_bundle.search_paths += [current_search_path]
```
Comment (charles-cooper) [inline vyper/semantics/analysis/imports.py:227]:
```suggestion
            self.input_bundle.search_paths.append(current_search_path)
```

Comment (charles-cooper) [reply vyper/semantics/analysis/imports.py:227]: this is kind of weird actually because on line 221, we already set search_paths to `[current_search_path]`

Comment (sandbubbles) [reply vyper/semantics/analysis/imports.py:227]: I think the issue is we need to set it temporarily to only one path, but then we need to add it to search paths in case there is an absolute import that uses it

**Comment by sandbubbles** [comment]
This should be ready for rereview

File: vyper/semantics/analysis/imports.py [hunk 10]
```diff
@@ -219,6 +214,20 @@ def _load_import_helper(
         search_paths = self.input_bundle.search_paths.copy()  # noqa: F841
         raise ModuleNotFound(module_str, hint=hint) from err
 
+    def _load_file(self, path: PathLike, level: int) -> CompilerInput:
+        ast = self.graph.current_module
+        current_search_path = Path(ast.resolved_path).parent
+
+        search_paths = self.absolute_search_paths if level == 0 else [current_search_path]
```
Comment (charles-cooper) [inline vyper/semantics/analysis/imports.py:221]: let's add a variable here for clarity:
```
        is_relative_import = (level != 0)
```

File: vyper/semantics/analysis/imports.py (hunk 9 above)
Comment (charles-cooper) [reply vyper/semantics/analysis/imports.py]: isn't that handled in the recursion by line 221?
https://github.com/vyperlang/vyper/pull/4268/files/c91f13e870bce8001eeec83dd96b8320c53d4343#diff-dc7dbf75948a3fce0db2cfa0fa819c964a534604d06f72ae88d18e3e3eef78b6R221

Comment (charles-cooper) [reply vyper/semantics/analysis/imports.py]: i pushed a clean up in a140226e89d803fdf3d254b3400f2af165238134

File: tests/unit/cli/vyper_json/test_compile_json.py (hunk 5 above)
Comment (charles-cooper) [reply tests/unit/cli/vyper_json/test_compile_json.py]: here it's preferred to use FileInput which already has the path/resolved_path metadata on it -- i've pushed it here 3359eaed90c89b52a525f68888be4ad14644630e

File: tests/functional/syntax/test_import.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+    lib1 = """
+def foo():
+    pass
+    """
+
+    input_bundle = make_input_bundle(
+        {"top.vy": top, "subdir0/lib0.vy": lib0, "subdir0/subdir1/lib1.vy": lib1}
+    )
+
+    with pytest.raises(ModuleNotFound):
```
Comment (charles-cooper) [inline tests/functional/syntax/test_import.py:30]: i think these might not be raising for the right reason -- if resolved_path is not set in compile_code, ModuleNotFound will be raised anyways

File: vyper/semantics/analysis/imports.py (hunk 9 above)
Comment (sandbubbles) [reply vyper/semantics/analysis/imports.py]: I though that example such as the following should work (but checked with python and you are right, it shouldn't).
```
#a.py
from .subdir import b
```

and 

```
#b.py
from subdir.subsubdir import c
```

**Review by cyberthirst — APPROVED** [review]

**Review by charles-cooper — APPROVED** [review]
lgtm, thank you
